json-el/src/main/scala/io/zeebe/msgpack/el/JsonCondition.scala

```diff
@@ -17,58 +17,43 @@ package io.zeebe.msgpack.el
 
 import org.agrona.DirectBuffer
 
-import scala.collection.mutable
-
-abstract class Comparison extends JsonCondition {
+trait Comparison {
   val x: JsonObject
   val y: JsonObject
 
-  def variableNames: mutable.HashSet[DirectBuffer] = {
-    val names = mutable.HashSet[DirectBuffer]()
-
-    if (x.isInstanceOf[JsonPath]) {
-      val path = x.asInstanceOf[JsonPath]
-      names += path.variableName;
-    }
-
-    if (y.isInstanceOf[JsonPath]) {
-      val path = y.asInstanceOf[JsonPath]
-      names += path.variableName;
+  val variableNames = {
+    (x, y) match {
+      case (x: JsonPath, y: JsonPath) => Set(x.variableName, y.variableName)
+      case (x: JsonPath, _)           => Set(x.variableName)
+      case (_, y: JsonPath)           => Set(y.variableName)
+      case _                          => Set[DirectBuffer]()
     }
-
-    return names
   }
 }
 
-abstract class Operator extends JsonCondition {
+trait Operator {
   val x: JsonCondition
   val y: JsonCondition
 
-  def variableNames: mutable.HashSet[DirectBuffer] = {
-    var set: mutable.HashSet[DirectBuffer] = mutable.HashSet()
-    set ++= x.variableNames
-    set ++= y.variableNames
-    return set
-  }
+  val variableNames = x.variableNames ++ y.variableNames
 }
 
 sealed trait JsonCondition {
-  def variableNames: mutable.HashSet[DirectBuffer]
+  val variableNames: Set[DirectBuffer]
 }
 
-case class Equal(x: JsonObject, y: JsonObject) extends Comparison
-
-case class NotEqual(x: JsonObject, y: JsonObject) extends Comparison
+case class Equal(x: JsonObject, y: JsonObject) extends JsonCondition with Comparison
 
-case class LessThan(x: JsonObject, y: JsonObject) extends Comparison
+case class NotEqual(x: JsonObject, y: JsonObject) extends JsonCondition with Comparison
 
-case class LessOrEqual(x: JsonObject, y: JsonObject) extends Comparison
+case class LessThan(x: JsonObject, y: JsonObject) extends JsonCondition with Comparison
```

Comment:
Why is Comparison and Operator not a JsonCondition anymore?

Reply:
Would be also possible. It is just more common that classes extend a sealed trait directly.

Reply:
Ok :)